File: day04/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

// Directions for 8 neighbors (including diagonals)
const int directions[8][2] = {
    {-1, -1}, {-1, 0}, {-1, 1},
    {0, -1},           {0, 1},
    {1, -1},  {1, 0},  {1, 1}
};

// Count adjacent rolls for a given position
int countAdjacentRolls(const vector<string>& grid, int r, int c) {
    int rows = grid.size();
    int cols = grid[0].size();
    int count = 0;

    for (int i = 0; i < 8; i++) {
        int nr = r + directions[i][0];
        int nc = c + directions[i][1];

        // Check bounds
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            if (grid[nr][nc] == '@') {
                count++;
            }
        }
    }

    return count;
}

int part1(const vector<string>& lines) {
    int rows = lines.size();
    int cols = lines[0].size();
    int accessible_count = 0;

    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (lines[r][c] == '@') {
                int adjacent = countAdjacentRolls(lines, r, c);
                if (adjacent < 4) {
                    accessible_count++;
                }
            }
        }
    }

    return accessible_count;
}

int part2(const vector<string>& lines) {
    // Create a mutable copy of the grid
    vector<string> grid = lines;
    int rows = grid.size();
    int cols = grid[0].size();
    int total_removed = 0;

    while (true) {
        // Find all rolls that can be removed in this iteration
        vector<pair<int, int>> removable;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == '@') {
                    int adjacent = countAdjacentRolls(grid, r, c);
                    if (adjacent < 4) {
                        removable.push_back({r, c});
                    }
                }
            }
        }

        // If no rolls can be removed, we're done
        if (removable.empty()) {
            break;
        }

        // Remove all accessible rolls
        for (const auto& pos : removable) {
            grid[pos.first][pos.second] = '.';
        }

        total_removed += removable.size();
    }

    return total_removed;
}

int main() {
    // Read input file
    ifstream file("../input.txt");
    if (!file.is_open()) {
        cerr << "Error: Could not open input file" << endl;
        return 1;
    }

    vector<string> lines;
    string line;
    while (getline(file, line)) {
        if (!line.empty()) {
            lines.push_back(line);
        }
    }
    file.close();

    if (lines.empty()) {
        cerr << "Error: No input data" << endl;
        return 1;
    }

    cout << "Part 1: " << part1(lines) << endl;
    cout << "Part 2: " << part2(lines) << endl;

    return 0;
}

File: day04/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use Cwd 'abs_path';

# Read input file
my $script_dir = dirname(abs_path($0));
my $input_file = "$script_dir/../input.txt";
open(my $fh, '<', $input_file) or die "Cannot open $input_file: $!";
my @lines = <$fh>;
close($fh);
chomp @lines;

# Parse grid
my @grid;
foreach my $line (@lines) {
    push @grid, [split(//, $line)];
}

my $rows = scalar @grid;
my $cols = scalar @{$grid[0]};

# Directions for 8 neighbors (including diagonals)
my @directions = (
    [-1, -1], [-1, 0], [-1, 1],
    [0, -1],           [0, 1],
    [1, -1],  [1, 0],  [1, 1]
);

# Count neighbors that are '@'
sub count_neighbors {
    my ($grid_ref, $r, $c) = @_;
    my $count = 0;

    foreach my $dir (@directions) {
        my $nr = $r + $dir->[0];
        my $nc = $c + $dir->[1];

        if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
            if ($grid_ref->[$nr][$nc] eq '@') {
                $count++;
            }
        }
    }

    return $count;
}

# Part 1: Count rolls with fewer than 4 adjacent rolls
sub part1 {
    my ($grid_ref) = @_;
    my $accessible_count = 0;

    for (my $r = 0; $r < $rows; $r++) {
        for (my $c = 0; $c < $cols; $c++) {
            if ($grid_ref->[$r][$c] eq '@') {
                my $neighbors = count_neighbors($grid_ref, $r, $c);
                if ($neighbors < 4) {
                    $accessible_count++;
                }
            }
        }
    }

    return $accessible_count;
}

# Part 2: Repeatedly remove accessible rolls until none remain
sub part2 {
    my ($grid_ref) = @_;

    # Deep copy the grid
    my @grid_copy;
    for (my $i = 0; $i < $rows; $i++) {
        push @grid_copy, [@{$grid_ref->[$i]}];
    }

    my $total_removed = 0;

    while (1) {
        # Find all accessible rolls
        my @accessible;

        for (my $r = 0; $r < $rows; $r++) {
            for (my $c = 0; $c < $cols; $c++) {
                if ($grid_copy[$r][$c] eq '@') {
                    my $neighbors = count_neighbors(\@grid_copy, $r, $c);
                    if ($neighbors < 4) {
                        push @accessible, [$r, $c];
                    }
                }
            }
        }

        # If no accessible rolls found, we're done
        if (scalar @accessible == 0) {
            last;
        }

        # Remove all accessible rolls
        foreach my $pos (@accessible) {
            my ($r, $c) = @$pos;
            $grid_copy[$r][$c] = '.';
            $total_removed++;
        }
    }

    return $total_removed;
}

# Run both parts
my $answer1 = part1(\@grid);
my $answer2 = part2(\@grid);

print "Part 1: $answer1\n";
print "Part 2: $answer2\n";
